rero_ils/modules/utils.py

```diff
@@ -1192,3 +1194,44 @@ def draw_row_content(row, pad=' '):
         ])+"\n"
 
     return table_header() + table_rows() + table_footer()
+
+
+def password_validator(pw, length=8, special_char=False):
+    """Validate the password.
+
+    :param pw: The password to validate.
+    :param length: Minimum password size.
+    :param special_char: If True add punctuation string
+    :return True if the password is valid.
+    """
+    return bool(
+        len(pw) >= length
+        and len(set(string.ascii_lowercase).intersection(pw))
+        and len(set(string.ascii_uppercase).intersection(pw))
+        and len(set(string.digits).intersection(pw))
+        and (not special_char or len(set(string.punctuation).intersection(pw)))
+    )
+
+
+def password_generator(length=8, special_char=False):
+    """Generate a password.
+
+    :param length: Minimum password size.
+    :param special_char: If True add punctuation string
+    :return the generated password.
+    """
+    min_length = 4 if special_char else 3
+    if length < min_length:
+        raise ValueError(f'Minimal size {min_length}')
+
+    password = [random.choice(string.ascii_lowercase)]
+    password.append(random.choice(string.ascii_uppercase))
+    password.append(random.choice(string.digits))
+    if special_char:
+        password.append(random.choice(string.punctuation))
+
+    while len(password) < length:
+        password.append(random.choice(string.ascii_letters + string.digits))
+
+    random.shuffle(password)
+    return ''.join(password)
```

Comment:
I suggest this approach. Not "THE" approach but a way to ensure that the password is valid according the 'validate' function. (inspiration : https://geekflare.com/password-generator-python-code/)

```python
import secrets
import string

def password_generator(length=8, max_attemps=500):
    attemps = 0
    alphabet = string.ascii_letters + string.digits + string.punctuation
    while True: 
        pwd = ''
        for i in range(length):
            pwd += ''.join(secrets.choice(alphabet))
        print(f"[{attemps}] {pwd}")
        try:
            password_validate(pwd)
            return pwd
        except Exception:
            pass
        attemps += 1
        if attemps > max_attemps:
            raise Exception('max_attemps occurs')
```

Reply:
We have defined that a minimum of one lower case character, one upper case character and one number is required. With this approach to random, it may take a while to get a good string and maybe get an error.